### DISCUSSION TITLE
IOException error when writing modified Assembly

### DISCUSSION THREAD

**Discussion by GTink911** [jbevain/cecil#933] [Q&A] [answered]
I've successfully managed to read and modify an Assembly within my code. However, when I try to write this Assembly back to the original file with the AssemblyDefinition.Write(path) function, it immediately crashes with an IOException. Here's my code:

```
using(var NormalAssembly = AssemblyDefinition.ReadAssembly(normal, new ReaderParameters { ReadWrite = true })){
        // Get the body of a function that is loaded very early on
        MethodDefinition method = NormalAssembly.MainModule.GetType("S_Video_Intro").Methods.Single(x => x.Name == "Awake");
        // Put a function call to our code at the start of that function
        method.Body.GetILProcessor().InsertBefore(method.Body.Instructions[0], Instruction.Create(OpCodes.Call, method.Module.ImportReference(hook)));
        // Put our changes back
        // TODO: Make this not crash if used by another program (though installer should prevent)
        NormalAssembly.Write(normal); // Crashes here

        Console.WriteLine("Normal injection complete."); // Never reaches here
}
```

And here's the text of the error:
```
An unhandled exception of type 'System.IO.IOException' occurred in System.Private.CoreLib.dll: 'The process cannot access the file '[FilePath]\Assembly-CSharp.dll' because it is being used by another process.'
```
Everything works until the NormalAssembly.Write() call, at which point that error occurs. The file path, type I'm accessing, etc are all valid. Additionally, I've used Process Explorer to look for all hooks to that DLL, and the only one that ever occurs is a single hook from this process. The process that uses the DLL is never running. This error occurs even immediately after restarting my entire PC.

I'm on version 0.11.5 of Mono.Cecil, and DotNet 8.0.201. 

Sorry in advance if I'm just missing something obvious, or if this is the wrong place to ask. I just can't find practically any documentation online for the library, and I can't locate anybody having the same problem that I haven't already covered.

Thanks in advance for the help, and let me know if there's any other information that would be helpful to provide.

**Comment by ltrzesniewski** [comment, marked as answer]
Since you're writing to the same path you read from, have you tried the overload without a parameter: `NormalAssembly.Write();`?

By default, the assembly file is kept open (probably due to the fact the `ReaderParameters.ReadingMode` property is set to `Deferred`).

**Reply by GTink911** [reply]
This worked, thanks! In the interest of avoiding needing to ask more, is there some public source of documentation that I just missed and can use in the future?

**Reply by ltrzesniewski** [reply]
Not as far as I know, sorry. Your best bet is to look at the source code or at other projects that use Cecil. 😕

**Reply by ltrzesniewski** [reply]
I just saw [this tweet](https://x.com/KooKiz/status/1760719331531297155?s=20) from @kevingosse which mentions [cecilifier.me](https://cecilifier.me) and thought you may be interested.
